`FeedView` optimizations

beardwin opened the pull request:
Came across your post on Reddit on this day of mourning. Thought i'd check out the project and was delighted to see it was RN. Thought i'd contribute a bit, if you're willing.

This PR does a few things:

1. It stabilizes the callbacks to FlashList (you usually want to stabilize virtualized list props as much as possible, as not doing it can chew performance)
2. Removes the inline functional component definitions (`footer`). This is a big one, as we were creating a new functional component on every render pass. Doing so prohibits React from optimizing the component, as we're blowing it away every time.
3. Cleans up the files a bit by creating some domain-specific components (`OverflowMenuButton`, `FeedSortButton`) for the Feed

Usually i would use Flipper and flamegraph the screen, but alas, it's not installed (yet?... hrmmmmm....). 

Lemme know if you have any questions - I'll post some comments inline here.

beardwin commented on components/ui/Feed/FeedOverflowButton.tsx:
```diff
@@ -0,0 +1,104 @@
+import { IconDots } from "tabler-icons-react-native";
```
Given we have a centralized set of APis and state management, I figured we could just create our own `FeedOverflowButton` for the `FeedView`. This just moves the functionality that we had in the FeedView for the overflow menu(s) into its own component, making it easier to maintain and - hopefully in due time - test.

beardwin commented on components/ui/Feed/FeedSortButton.tsx:
```diff
@@ -0,0 +1,62 @@
+import { useTheme } from "native-base";
```
Same philosophy as the FeedOverflowMenuButton. No need to set state in the FeedView for the sort button, as it's all based on the `SortType` from the Lemmy api anyways. Now we can just pass it in, and the component can render whatever component it wants.

beardwin commented on components/ui/Feed/FeedSortButton.tsx:
```diff
@@ -0,0 +1,62 @@
+      }
+    );
+  };
+  return <HeaderIconButton icon={SortIconType[feed.sort]} onPress={onPress} />;
```
This is admittedly dangerous, but is basically the same logic we had before. Dangerous because `SortIconType` doesn't contain definitions for all the different `SortType`s, so it's possible for future selves to update sortOptions but not update SortIconType, which would result in... whatever happens when you try to wrap a `Pressable` around an `undefined`

sgriff96 replied:
We can update some typescript stuff to help prevent that

beardwin commented on components/ui/Feed/FeedView.tsx:
```diff
@@ -80,156 +54,96 @@ function FeedView({ feed, community = false, header }: FeedViewProps) {
-    ];
-    const cancelButtonIndex = 6;
+  const postCommunity: Community = useMemo(() => {
+    if (!firstPost || !community) return undefined;
```
instead of using a side effect to update 2 refs for maintaining the community ID and community Name, we can derive this information from the first post and the `community` flag being passed in. Doing this allows us to update the header as part of a normal render pass, if necessary.

beardwin commented on components/ui/Feed/FeedView.tsx:
```diff
@@ -80,156 +54,96 @@ function FeedView({ feed, community = false, header }: FeedViewProps) {
-    } else {
-      const options = ["All", "Local", "Subscribed", "Cancel"];
-      const cancelButtonIndex = 3;
+  const renderItem = React.useCallback(
```
memoizing a few of the callbacks we're passing to FlashList; you typically don't want anonymous functions being passed to a list. It's notorious for impacting performance.

beardwin commented on components/ui/Feed/FeedView.tsx:
```diff
@@ -238,27 +152,6 @@ function FeedView({ feed, community = false, header }: FeedViewProps) {
-    return null;
-  };
-
-  const HeaderComponent = header;
```
This seemed unnecessary, and is in fact kinda weird given `header` is completely optional and can be `undefined`. But i could be wrong. Either way, i think it's more straight forward to just render it out via `{header}` instead of abstracting it into the `HeaderComponent`.

beardwin commented on components/ui/Feed/FeedView.tsx:
```diff
@@ -238,27 +152,6 @@ function FeedView({ feed, community = false, header }: FeedViewProps) {
     />
   );
 
-  const footer = () => {
```
This is a big no-no in react. This _defines_ the functional component `footer` on every single render pass. As a result, we're always changing the definition of the component, and thus React can't optimize it.

I moved it into `Footer.tsx`, below.

beardwin commented on components/ui/Feed/FeedView.tsx:
```diff
@@ -293,6 +195,8 @@ function FeedView({ feed, community = false, header }: FeedViewProps) {
   );
 }
 
+const keyExtractor = (item: PostView) => item.post.id.toString();
```
moved this out of the component as it's a pure function, and is not stable. The alternative is to wrap it in `useCallback`, but if we don't need to it's best not to.

beardwin commented on components/ui/Feed/FeedView.tsx:
```diff
@@ -1,74 +1,48 @@
   // State Props
   // TODO Handle this
   // eslint-disable-next-line @typescript-eslint/no-unused-vars
   const [endReached, setEndReached] = useState(false);
```
I'm not familiar with the Lemmy API (honestly, I'd never heard of Lemmy until today), but ideally it would have some kind of pagination information. If that's the case, then we can derive from that instead of storing a `endReached` state.

gkasdorf commented:
Oh this is amazing. All things that I know we need to do and am hopefully finally going to find the time to do.

I'm also not generally a React dev, usually using Vue. So I'm totally happy whenever someone points out ways to improve optimization and follow better practices! I seriously appreciate it. 

We're going to start going through and refactoring things to clean up and optimize, so this is perfectly in line with what we are going to start working on. 

Once I get a chance I'll go over all of this and take note of the changes. Thank you!

beardwin commented:
> Oh this is amazing. All things that I know we need to do and am hopefully finally going to find the time to do.
> 
> I'm also not generally a React dev, usually using Vue. So I'm totally happy whenever someone points out ways to improve optimization and follow better practices! I seriously appreciate it.
> 
> We're going to start going through and refactoring things to clean up and optimize, so this is perfectly in line with what we are going to start working on.
> 
> Once I get a chance I'll go over all of this and take note of the changes. Thank you!

Rock and roll, my dude! I just rebased and incorporated the changes you made for the contextual overflow icon. 

I'm open to discussion about my changes. I've been doing RN fulltime at Shopify for 5 years, and there's still basics to this black magic which can trip me up. RN (and React in general) is deceptively hard. But hey, it's how we learn, and that is always fun.

Great initiative on this work, by the way. I might submit some other PRs. Sometimes I just like to hack (like today - day off yet.... here I am)

gkasdorf commented:
Maybe you can provide some insight (and it might be in this PR once I take a close look) into best practices on maintaining something like the image height for each of our items.

Currently, I'm storing that info whenever the item's props change so that when we scroll back up we can quickly set that data back without having to wait for FastImage's `onLoad()`. I assume this is the best (or maybe the only?) way to handle this.

Super awesome library by the way. Given that performance is already pretty great without these fixes here really goes to show the work (and the work of recyclerlistview's maintainers) that was put into this.

Thanks!

beardwin commented:
> Maybe you can provide some insight (and it might be in this PR once I take a close look) into best practices on maintaining something like the image height for each of our items.
> 
> Currently, I'm storing that info whenever the item's props change so that when we scroll back up we can quickly set that data back without having to wait for FastImage's `onLoad()`. I assume this is the best (or maybe the only?) way to handle this.
> 
> Super awesome library by the way. Given that performance is already pretty great without these fixes here really goes to show the work (and the work of recyclerlistview's maintainers) that was put into this.
> 
> Thanks!

I gotta give credit where credit is due, and I am not the recipient! ha! The dudes working on FlashList are brilliant, most def.

As for FastImage.... I'm pretty sure it caches the images, no? So the `onLoad` event that it fires should, in theory, be nearly immediate if the referenced image has already been downloaded. But I could be mistaken (someTimes it's hard to remember what we've done in house, what we've changed, vs what is au-natural).

I don't have anything in this PR about the images, but if you point me in the right direction i might be able to take a look at some point. I don't have a _lot_ of capacity for contributing regularly, but I can always try and help out a bit when I can.

gkasdorf commented:
The behavior I saw was that when scrolling back up, the new image gets replaced with the old one, which causes some pretty serious "push down" effects when the new image gets placed in the component. 

Since we are rendering images of various sizes and not using fixed dimensions for those images, I believe (I can't tell for sure since it is off screen) that the item's FastImage style is still set to have the height of the last image.

FastImage does cache the image, however, you are not able to get the height of that image until onLoad is called, so on the first render the height is outdated. On the second render the height is updated, causing the affect.

As a solution, this is what I am doing

```ts
  const lastPostId = useRef(postId);

  if (recycled && postId !== lastPostId.current) {
    recycled.current = {
      ...recycled.current,
      [lastPostId.current]: {
        height,
        width,
      },
    };

    if (recycled.current[postId]) {
      setHeight(recycled.current[postId].height);
      setWidth(recycled.current[postId].width);
    }

    lastPostId.current = postId;
  }
  ```

We check if recycled is set since the component does get reused in places where we don't care about recycling. Then, following documentation, we check the new post ID against the old one.

This resolves our issue of height (and we probably don't need to actually use state here looking at it again, but that's what I have in there currently.)

Here is a link to the code in question: https://github.com/gkasdorf/memmy/blob/157ecde242f9b048ff55a99937c84f0f1bc6ffaa/components/ui/image/MemoizedFastImage.tsx#L36

Seriously appreciate you taking the time to help. Cheers!

beardwin commented:
Ah yea, that's a tricky issue, for sure. I just went to check what Apollo does for image aspect ratios in their feed, but alas, I am too late. RIP.

I will noodle on this. FastImage _does_ have the ability of prefetching, and I'm wondering if that would be an approach we could take (prefetch the cached image and then use the dimensions)....that's purely off the top of my head, though. Could be way off.
